base/iterators.jl

```diff
@@ -1164,8 +1164,15 @@ julia> [(x,y) for x in 0:1 for y in 'a':'c']  # collects generators involving It
 """
 flatten(itr) = Flatten(itr)
 
-eltype(::Type{Flatten{I}}) where {I} = eltype(eltype(I))
 eltype(::Type{Flatten{Tuple{}}}) = eltype(Tuple{})
+eltype(::Type{Base.Iterators.Flatten{T}}) where T = tuple_eltype_join(T)
+
+tuple_eltype_join(::Type{Tuple{T}}) where T = eltype(T)
+function tuple_eltype_join(::Type{T}) where {T <: Tuple}
+    E = eltype(Base.tuple_type_head(T))
+    E === Any ? Any : typejoin(E, tuple_eltype_join(Base.tuple_type_tail(T)))
+end
```

Comment:
```suggestion
tuple_eltype_join(::Type{T}) where {T <: Tuple} = Base.typejoin(eltype.(fieldtypes(T))...)
```
Does the above also work, or are there reasons to use the recursive implementation? I suppose the recursive implementation might terminate faster if the typejoin becomes `Any`?

Reply:
(Superceded by next suggestion)